### ISSUE TITLE
Add support for linear gradients

### ISSUE THREAD

**Issue by mojavelinux** [mogest/prawn-svg#62]
Add support for the linear gradient fill.

Many diagramming tools output shapes using linear gradients. In order for these diagrams to look right in the PDF, prawn-svg will need to apply the requested linear gradients. I'm sure there are many other use cases for linear gradients, but this one seems to come up the most.

Here's one such example:

https://github.com/openEHR/spec-publish-asciidoc/blob/22f7f71115cfe1e826b6baf1f99ba127cd4be735/docs/AOM2/uml_diagrams/AOM-archetype.svg

Prawn does support linear gradients, so it should be possible in theory to delegate to prawn to handle this type of fill.

**Comment by mogest** [comment]
I've been working on gradients the last two nights, actually.

Prawn does support gradients, but very badly.  The co-ordinates you pass the fill_gradient method are absolute co-ordinates on the page, and not affected by any scaling, translation or rotation that's done in user space.

That's just how PDF works, although there are two options to get them into user space: XForm objects (apparently, I have not confirmed but the spec leads me to believe that's possible), and using a shading drawing instruction instead of stroke/fill.  The latter wouldn't give individual control over stroke and fill colors, so although some kludge might be able to be done using clips and drawing the stroke separately from the fill maybe.

To make it even harder, SVG allows (and by default accepts) x/y coordinates between 0 and 100% which are mapped to the bounding box of the shape being drawn, and PDF does not.  To support SVG gradients, I'd have to work out the bounding box for every shape that's able to be drawn.  Not impossible, but not trivial, and definitely requiring a substantial rewrite.

And although PDF does support multi-stage gradients like SVG, prawn doesn't.  I've seen a nice implementation of shading with opacity in nodejs-images-to-pdf, so again, possible, but work required.

In short: a lot more investigation needs to be done, let alone code.  I'll do it as I find time, but not promising anything here!

**Comment by mojavelinux** [comment]
Wow! The devil is always in the details, isn't it?

Hopefully, PDF will reveal some secrets that you can tap into. I'll try to grab some interest in this issue to see if it helps uncover more research. I'll try to help in anyway I can.

**Comment by mogest** [comment]
Progress update: I've prototyped a fix to prawn so it will render gradients correctly.  See https://github.com/prawnpdf/prawn/issues/891.  Waiting to see whether I get any comments about my approach before submitting a PR.

With that fixed, it'll be easier to implement them in prawn-svg.  Is going to be a pain having to calculate the bounding box for everything prawn-svg can draw, though!

**Comment by mojavelinux** [comment]
Fantastic! Thanks for the update. I'm following the progress with great anticipation.

**Comment by mogest** [comment]
Hey @mojavelinux, I haven't dealt with the new prawn maintainers before.  Are they likely to respond to the issue, or should I just go ahead with the PR?

**Comment by mojavelinux** [comment]
@mogest The new maintainers, and in particular Evan, are quite responsive and willing to learn what is necessary to address issues in Prawn. I'd recommend moving forward with the PR as that seems to get the conversation focused on the scale of the change.

I'm really excited about these improvements because it's refreshing part of the code that seems to have gone dormant.

**Comment by mogest** [comment]
Basic support implemented in 5c7fb8ab1c80cb862e9db765508490e80860585d!

I'll wait until Prawn 2.0.3 gets released (assuming the second PR gets accepted before then) and then I'll release a new prawn-svg.

**Comment by mojavelinux** [comment]
I'm really excited about this! It's going to open a lot of doors for document authors. I'm very grateful for your contributions to Prawn. Thanks!

**Comment by mogest** [comment]
Just an update: we're waiting on https://github.com/prawnpdf/prawn/pull/902 to go in and be released before this feature can go live.

**Comment by v3rmin** [comment]
According to the prawns changelog this should be possible to implement by now? All your pull requests have been merged?

**Comment by mogest** [comment]
@v3rmin, yeah, the PR that was required was merged on May 14, but there hasn't been a release since March 😞

**Comment by TobiasHector** [comment]
@mogest Prawn has been updated to 2.2.2 now - any chance of this progressing now we're a few versions further? 🤞

**Comment by mogest** [comment]
Ooo, yes, looks like it.  Let me play.

**Comment by mogest** [comment]
Very happy, at long last, to bring gradients to prawn-svg 😄 🎉 🎈 🎈 🎉
